### PR TITLE
4D Volume Rendering using the ArcGIS JS API external renderer API

### PR DESCRIPTION
**Pull request by eile**
[Live Version](https://eile.github.io/devsummit-2019-3D-jsapi/Advanced-WebGL-in-3D-SceneViews/ArcGIS-Volume-Renderer/)

* For the live version to work, I've checked in the generated js files. The alternative is to load a 6MB js file to transpile ts in the browser. This also does not work out of the box since I would need to patch in a tsconfig.
* The live version loads the dataset from git-lfs, so animation is really slow since a full time series is about 1GB
* The local version uses the local files

### CODE REVIEW

File: Advanced-WebGL-in-3D-SceneViews/ArcGIS-Volume-Renderer/app/Application.ts [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+      const aa = dataList as HTMLOptionElement;
+      renderer.data = aa.value;
+    };
+
+    window["app"] = this;
+    window["view"] = this.view;
+  }
+}
+
+const spatialReference = { wkid: 102100 };
```
Comment (joejohndoe) [inline Advanced-WebGL-in-3D-SceneViews/ArcGIS-Volume-Renderer/app/Application.ts]: would be a bit nicer to use `SpatialReference.WebMercator` in this case

File: Advanced-WebGL-in-3D-SceneViews/ArcGIS-Volume-Renderer/app/Application.ts [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        heading: 0,
+        tilt: 53
+      },
+      container: "view",
+      viewingMode: "local",
+      map: new Map({ basemap: "satellite", layers: [legendLayer] }),
+      environment: {
+        background: { type: "color", color: [255, 244, 252, 1] },
+        starsEnabled: false,
+        atmosphereEnabled: false // Atmosphere would overwrite volume rendering
```
Comment (joejohndoe) [inline Advanced-WebGL-in-3D-SceneViews/ArcGIS-Volume-Renderer/app/Application.ts]: this is actually a usability issue with external renderers we should investigate IMO...have you discussed it with Adrian or Jesse? perhaps a ticket.

File: Advanced-WebGL-in-3D-SceneViews/ArcGIS-Volume-Renderer/app/Application.ts [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+      {
+        layer: legendLayer,
+        title:
+          "Hurricane Isabel data produced by the Weather Research and Forecast (WRF) model, courtesy of NCAR, and the U.S. National Science Foundation (NSF)"
+      }
+    ];
+
+    this.view.ui.add(legend, "top-right"); // Add legend to the bottom right corner of the view
+    this.view.ui.add("infoDiv", "top-right"); // Data set list to top right
+    this.view["renderContext"] = "webgl2"; // Needed for 3D textures
```
Comment (joejohndoe) [inline Advanced-WebGL-in-3D-SceneViews/ArcGIS-Volume-Renderer/app/Application.ts]: Extend comment to state that this uses unsupported, internal functionality

File: Advanced-WebGL-in-3D-SceneViews/ArcGIS-Volume-Renderer/app/VolumeRenderer.ts [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+};
+
+const localURL = "../isabeldata/";
+const remoteURL =
+  "https://media.githubusercontent.com/media/eile/devsummit-2019-3D-jsapi/master/Advanced-WebGL-in-3D-SceneViews/isabeldata/";
+let dataSetURL = localURL;
+
+// Set up typings
+type RenderContext = __esri.RenderContext;
+type ExternalRenderer = __esri.ExternalRenderer;
```
Comment (joejohndoe) [inline Advanced-WebGL-in-3D-SceneViews/ArcGIS-Volume-Renderer/app/VolumeRenderer.ts]: why is this needed?

File: Advanced-WebGL-in-3D-SceneViews/ArcGIS-Volume-Renderer/app/VolumeRenderer.ts [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+        }
+      }
+    );
+  }
+
+  private _updateLegend(): void {
+    const image = new Image();
+    image.src = dataSetURL + this._data[0] + ".png";
+    image.crossOrigin = "Anonymous";
+    image.onload = () => {
```
Comment (joejohndoe) [inline Advanced-WebGL-in-3D-SceneViews/ArcGIS-Volume-Renderer/app/VolumeRenderer.ts]: could you use `esri/request` for this (`responseType: "image"`)?

also, a quick comment about what's going on here might be helpful.

File: Advanced-WebGL-in-3D-SceneViews/ArcGIS-Volume-Renderer/app/VolumeRenderer.ts [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+      for (let i = 0; i < pixels.length; i += 4) {
+        stops[i / 4] = {
+          value: (min + i * valueStep).toFixed(1),
+          label: "",
+          color: new EsriColor([pixels[i + 0], pixels[i + 1], pixels[i + 2], 1])
+        };
+      }
+
+      stops[0].label = min.toFixed(1) + " " + this._data[3];
+      stops[stops.length - 1].label = max.toFixed(1) + " " + this._data[3];
```
Comment (joejohndoe) [inline Advanced-WebGL-in-3D-SceneViews/ArcGIS-Volume-Renderer/app/VolumeRenderer.ts]: would be more posh to use template literals :)

**Comment by eile** [comment]
> sourcemap files can be removed

I thought they might be useful so people can step through the .ts code?

File: Advanced-WebGL-in-3D-SceneViews/ArcGIS-Volume-Renderer/app/Application.ts (hunk 1 above)
Comment (eile) [reply Advanced-WebGL-in-3D-SceneViews/ArcGIS-Volume-Renderer/app/Application.ts]: done

File: Advanced-WebGL-in-3D-SceneViews/ArcGIS-Volume-Renderer/app/Application.ts (hunk 3 above)
Comment (eile) [reply Advanced-WebGL-in-3D-SceneViews/ArcGIS-Volume-Renderer/app/Application.ts]: done

File: Advanced-WebGL-in-3D-SceneViews/ArcGIS-Volume-Renderer/app/VolumeRenderer.ts (hunk 4 above)
Comment (eile) [reply Advanced-WebGL-in-3D-SceneViews/ArcGIS-Volume-Renderer/app/VolumeRenderer.ts]: Goes back to the slack chat: If I understand correctly, these two types are missing from the public typings. See internal PR 19451.

File: Advanced-WebGL-in-3D-SceneViews/ArcGIS-Volume-Renderer/app/VolumeRenderer.ts (hunk 5 above)
Comment (eile) [reply Advanced-WebGL-in-3D-SceneViews/ArcGIS-Volume-Renderer/app/VolumeRenderer.ts]: Will do

**Comment by joejohndoe** [comment]
>> sourcemap files can be removed

> I thought they might be useful so people can step through the .ts code?

I guess they could be. Up to you 👍

File: Advanced-WebGL-in-3D-SceneViews/ArcGIS-Volume-Renderer/app/VolumeRenderer.ts (hunk 6 above)
Comment (eile) [reply Advanced-WebGL-in-3D-SceneViews/ArcGIS-Volume-Renderer/app/VolumeRenderer.ts]: done

File: Advanced-WebGL-in-3D-SceneViews/ArcGIS-Volume-Renderer/app/VolumeRenderer.ts (hunk 4 above)
Comment (joejohndoe) [reply Advanced-WebGL-in-3D-SceneViews/ArcGIS-Volume-Renderer/app/VolumeRenderer.ts]: all clear, let's update once that issue is resolved. you could change the comment to reflect that this is fixing a typings issue if that takes longer than expected.

**Comment by eile** [comment]
@arnofiva FYI
